Add driver support for ADC1 of stm32

cybertale opened the pull request:
This PR contains pinmux, dts and driver support for ADC1 for all series of stm32 that zephyr supports.

cybertale commented:
To test this code, we should enable adc driver in menuconfig and add pinmux config in board specific file.
For f4 board, my configuration is 
```
#ifdef CONFIG_ADC0_CHAN0
	{STM32_PIN_PA0, STM32F4_PINMUX_FUNC_PA0_ADC123_IN0},
#endif
```

Test application code is
```/*
 * Copyright (c) 2012-2014 Wind River Systems, Inc.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include <adc.h>
#include <zephyr.h>
#include <misc/printk.h>

uint16_t temp_data[8];

static const struct adc_channel_cfg channel_cfg_0 = {
        .channel_id       = 0,
};

const struct adc_sequence sequence = {
        .channels    = BIT(0),
        .buffer      = temp_data,
        .buffer_size = sizeof(temp_data),
        .resolution  = 12,
};

void main(void)
{
        int ret;
        struct device *adc_dev = device_get_binding(DT_ADC_0_NAME);
        adc_channel_setup(adc_dev, &channel_cfg_0);

        while (1) {
                ret = adc_read(adc_dev, &sequence);
                if (ret != 0) {
                        printk("adc_read() failed with code %d", ret);
                } else {
                        printk("adc_read() value %d\n", temp_data[0]);
                }
                k_sleep(1000);
        }
}
```

cybertale commented:
Hi @erwango , I'v only test this code on f4 board, could you test it on other boards that available for you when you got some time?

cybertale commented:
Generic DMA driver of stm32 is under development in #13364 , without DMA support many stm32s won't be able to catch up EOC interrupts generated by channels conversion, this will cause overrun error and lost data. So Only one channel can be put in one adc sequence in this driver.

KwonTae-young commented:
I was interested in the ADC driver for STM32.
I tested it on [ST STM32F4DISCOVERY](https://docs.zephyrproject.org/latest/boards/arm/stm32f4_disco/doc/index.html) as shown below.
``` bash
west init zephyrproject -m https://github.com/KwonTae-young/zephyr/ --mr adc_stm32-new_stm32f4-disco
cd zephyrproject/zephyr
source zephyr-env.sh
mkdir samples/basic/adc/build
cmake -DBOARD=stm32f4_disco ..
make -j8 flash
```

ADC works well.
```
********** Booting Zephyr OS zephyr-v1.13.0-4846-g2eee90a41b *****
adc_read() value 1128
adc_read() value 952
adc_read() value 927
adc_read() value 1083
adc_read() value 916
adc_read() value 904
```

cybertale commented:
@KwonTae-young Thanks for your interest. :grin:

anangl commented on drivers/adc/adc_stm32.c:
```diff
@@ -0,0 +1,288 @@
+
+	adc_context_lock(&data->ctx, false, NULL);
+
+	return start_read(dev, sequence);
```
You need to correct the locking scheme. Currently if an invalid sequence is requested (with an invalid resolution, for instance), the driver will remain locked after reporting the error.
See for instance how it is done in [adc_nrfx_adc](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/adc/adc_nrfx_adc.c#L214) (the [recent commit](https://github.com/zephyrproject-rtos/zephyr/commit/0906a51dacc947c5fa19a80e208549f9039556cd#diff-f8d824106434c62f9e6458df9242c4f5) that modified this driver contains a similar correction to what you need to apply) or in [adc_mcux_adc16](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/adc/adc_mcux_adc16.c#L150) that looks quite similar to your driver.

There is a test for ADC drivers in the Zephyr source tree that will help you check if the API implementation is correct. See [tests/drivers/adc/adc_api](https://github.com/zephyrproject-rtos/zephyr/tree/master/tests/drivers/adc/adc_api).

anangl commented on drivers/adc/adc_stm32.c:
```diff
@@ -0,0 +1,288 @@
+		(struct adc_stm32_cfg *)dev->config->config_info;
+	ADC_TypeDef *adc = config->base;
+
+	*data->buffer = adc->DR;
```
The buffer pointer is not increased, so all the values read in a given sequence will be stored in the same place. Hence, the driver will work correctly only for simplest measurements (single channel, single sampling).
Again, try to run the [adc_api test](https://github.com/zephyrproject-rtos/zephyr/tree/master/tests/drivers/adc/adc_api) to check if your driver behaves as expected.

cybertale replied:
Hi @anangl , this is because currently we do not have a generic dma driver for stm32 and without dma, many stm32s won't be able to catch up every channel's EOC interrupt of a sequence, and overrun detection alerts. 
I'm working on dma driver on #13364, so before it's done, that's why only simple conversion is supported in this driver.
I'll try adc_api test, thank you.

cybertale replied:
Sorry, I found this should be increased even in my situation during running the adc_api test, I'll change this.

anangl commented on drivers/adc/Kconfig.stm32:
```diff
@@ -0,0 +1,155 @@
+	menu ADC_0
+		config ADC0_CHAN0
+			bool "channel 0 (PA0)"
+			default n
```
`default n` is redundant.

erwango commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -293,6 +293,15 @@
+		adc1: adc@40012400 {
+			compatible = "st,stm32-adc";
+			reg = <0x40012400 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000200>;
```
STM32_CLOCK_BUS_APB1_2

erwango commented on drivers/adc/Kconfig.stm32:
```diff
@@ -0,0 +1,155 @@
+
+if ADC_STM32
+
+	menu ADC_0
```
Rename to ADC1

cybertale replied:
I really would like to do this while kconfig system of adc has a minor problem against this, which I described in #13501 . I will change them all to ADC_1 but then this should be merged after that PR.

erwango commented on drivers/adc/Kconfig.stm32:
```diff
@@ -0,0 +1,155 @@
+		config ADC0_CHAN1
+			bool "channel 1 (PA1)"
+			default n
+			depends on ADC_0
```
Put whole section under "if ADC_1" and remove "depends on ADC_1"

erwango commented on drivers/adc/adc_stm32_v2.c:
```diff
@@ -0,0 +1,50 @@
+
+void adc_stm32_init_soc(ADC_TypeDef *adc)
+{
+	adc->CR2 |= ADC_CR2_ADON;
```
Why not using `LL_ADC_Enable` and factorize this part?

cybertale replied:
LL_ADC_Enable can be applied here, but I was thinking if I used LL lib somewhere, it's better to use LL lib everywhere. We talked about why I do not think LL lib is good for this generic driver, and except this function, some functions like getting EOC flags from registers uses different functions between different series of stm32 in LL lib.
So should I use LL and direct register access in the same driver?
I personally prefer to use direct register access for all operations.

erwango reviewed (changes requested):
Initial feedback

cybertale commented:
This PR should not be only merged until #13501 is merged.

anangl commented on drivers/adc/adc_stm32.c:
```diff
@@ -0,0 +1,407 @@
+
+	adc_context_start_read(&data->ctx, sequence);
+	err = adc_context_wait_for_completion(&data->ctx);
+	adc_context_release(&data->ctx, err);
```
Remove this line. The lock is now released in `adc_stm32_read[_sync]`.

cybertale replied:
Right.

cybertale commented:
Updated to use LL lib.

erwango commented on CODEOWNERS:
```diff
@@ -131,6 +131,7 @@
 /dts/arm/st/                              @erwango
 /dts/arm/nordic/                          @ioannisg @carlescufi
 /dts/bindings/can                         @alexanderwachter
+/dts/bindings/iio/adc/st,stm32-adc.yaml   @cybertale
```
Please also add drivers/adc/adc_stm32.c

cybertale replied:
OK

erwango commented:
@cybertale , since your work is close to mature and complete, could you update the description of the commit with a description of the work and the way it was tested? (this could edited later on with further addition)

cybertale commented:
@erwango Sure!

anangl commented on tests/drivers/adc/adc_api/src/test_adc.c:
```diff
@@ -121,6 +121,78 @@
+#define ADC_GAIN		ADC_GAIN_1
+#define ADC_REFERENCE		ADC_REF_INTERNAL
+#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
+#define ADC_1ST_CHANNEL_ID	0
```
These parameters are exactly the same as the ones for `NUCLEO_F091RC`. Wouldn't it be better to use:
```
#elif defined(CONFIG_BOARD_NUCLEO_F091RC) || \
      defined(CONFIG_BOARD_NUCLEO_F103RB)
```

Same applies to a couple of Nucleo boards using 10-bit resolution.

cybertale replied:
Yes, that's better

anangl commented on tests/drivers/adc/adc_api/src/test_adc.c:
```diff
@@ -121,6 +121,78 @@
+#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
+#define ADC_1ST_CHANNEL_ID	0
+
+#elif defined(CONFIG_BOARD_NUCLEO_F103RB)
```
Duplicate entry.

dbkinder reviewed (approved):
+1 for doc changes

erwango commented on drivers/adc/Kconfig.stm32:
```diff
@@ -0,0 +1,25 @@
+	default y
+	help
+		Enable ADC1
+
```
Miss ADC_2 and _3, since available in the driver

cybertale replied:
I'm actually thinking about removing them, because there are so many different product lines, and different amount of ADCs they have, and also some ADCs share interrupt vector, so there should be some code for identifying which instance calls the interrupt. I would like another PR to do these things because this PR can be merged so users of STM32 can have basic ADC support.
But I'm worrying what Geo said about the merge window, but I see zephyr's doc says

> As a general rule, if you miss the merge window for a given feature, the best thing to do is to wait for the next development cycle. (An occasional exception is made for drivers for previously unsupported hardware; if they do not touch any other in-tree code, they cannot cause regressions and should be safe to add at any time).

So if this passes all tests and clean, we should be able to merge this PR soon, right?

erwango replied:
This is up to @galak to settle on this. Though, I'm not quite in favor of it. Even it could not be a source of regression, it could be a source of bugs, and branch maturity is tightly controlled.
Besides, there are several PR that could pretend be merged as this PR (L1/G0/WB). There is no reason this one is privileged. 
So I think the paragraph you mention should actually be corrected.

cybertale replied:
Understand, so I'll just wait. ;)

erwango reviewed (changes requested):
This is promising, some more comments

cybertale commented:
It seems like the Shippable tests failed because another work is under going, #14242 should be able to solve the problem.

anangl commented on boards/arm/nucleo_f401re/pinmux.c:
```diff
@@ -41,6 +41,8 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PB15, STM32F4_PINMUX_FUNC_PB15_SPI2_MOSI},
 #endif	/* CONFIG_SPI_1 */
+#ifdef CONFIG_ADC_1
+#endif	/* CONFIG_ADC_1 */
```
Is this empty `#ifdef` block intentional?

cybertale replied:
Sorry, my mistake

GeorgeCGV commented:
@cybertale, tested on L452 SoC, works great.

shphilippe commented on drivers/adc/adc_stm32.c:
```diff
@@ -0,0 +1,588 @@
+	LL_ADC_EnableInternalRegulator(adc);
+	u32_t wait_loop_index = ((LL_ADC_DELAY_INTERNAL_REGUL_STAB_US / 10UL) *
+				(SystemCoreClock / (100000UL * 2UL)));
+	while (wait_loop_index != 0UL) {
```
k_busy_wait(LL_ADC_DELAY_INTERNAL_REGUL_STAB_US);

cybertale replied:
That's better.

shphilippe commented:
btw, I've added support for the STM32G0 in this branch: https://github.com/shphilippe/zephyr/tree/stm32g070-adc

cybertale commented:
> btw, I've added support for the STM32G0 in this branch: https://github.com/shphilippe/zephyr/tree/stm32g070-adc

I see G0 support is not in plan of milestone v1.15, right? So you can feel free to push support for ADC of G0 after this. Or what's your plan? :)

shphilippe commented:
I don't have much plans. Just trying to contribute back code which can be open sourced :)

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -0,0 +1,584 @@
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
```
Break a line here

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -0,0 +1,584 @@
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+				sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
```
Same here

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -0,0 +1,584 @@
+	LL_ADC_EnableIT_EOCS(adc);
+#endif
+
+	adc_context_start_read(&data->ctx, sequence);
```
Break a line here.

erwango reviewed (commented):
Mostly nitpicks

cybertale commented:
Hi @erwango , I added some comments and fixed style issues, how about this time?

cybertale committed:
dts: arm: stm32: add dts support for ADC1 of stm32

All series of stm32 have at least one ADC instance and this commit adds
one ADC node to the root dts file of each soc, and also adds fixing up
mappings to them.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: pinmux: stm32: add pinmux support for ADC of stm32

This commit adds pinmux defines for all the external ADC lines
supported by stm32. All defines are named after the datasheet of the
corresponding product lines.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: adc: add driver support for ADC1 of stm32

This commit adds driver support for ADC1 on all 8 supported series of
stm32 with resolution and conversion time selection and calibration.

Currently DMA is not supported for all series, and without it, zephyr
won't be able to catch up ADC's end of conversion interrupt, so this
version of the driver supports one channel conversion only. Users want
multi-channel conversion should use multiple sequences in their app
code.

This driver uses LL lib rather than HAL because the current HAL lib for
ADC will call HAL_DMA_* functions rather than using zephyr's common DMA
interface, so that way the driver will break the consistency of the
code.

This driver has been tested on multiple nucleo boards including
NUCLEO_F091RC/F103RB/F207ZG/F302R8/F401RE/F746ZG/L073RZ/L476RG and all
passed the test cases in tests/drivers/adc/adc_api. If the external ADC
line is floating, it may fail the tests since ADC may get 0V and the
test cases think 0 is failing. Connect it to any voltage source between
0-3.3V will help passing the test cases.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
test: driver: adc: add some nucleo boards support for adc test cases

Some configuration for the boards have to be added into test_adc.c file
so user can test driver with the test cases. Several nucleo boards are
added including F091RC/F103RB/F207ZG/F302R8/F401RE/F746ZG/L073RZ/L476RG.

And also ADC dts and pinmux configuration are added into boards own
pinmux.c and dts file.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
doc: boards: arm: nucleo: add ADC support to the boards doc

The ADC driver in this PR has been tested working on these nucleo
boards, so ADC support is added to the boards doc.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

Odjek commented:
Hello. I tested the driver on stm32f413zh. It works well on PA0 but when I try to change the ADC pin it's not working anymore. Any ideas ? Thanks.
